Tests/Engine/InvokeScriptAnalyzer.tests.ps1

```diff
@@ -397,6 +397,18 @@ Describe "Test CustomizedRulePath" {
                 }
             }
 
+            It "Should process relative settings path even when settings path object is an expression" {
+                try {
+                    $initialLocation = Get-Location
+                    Set-Location $PSScriptRoot
+                    $warnings = Invoke-ScriptAnalyzer -ScriptDefinition 'gci' -Settings (Join-Path (Get-Location).Path '.\SettingsTest\..\SettingsTest\Project1\PSScriptAnalyzerSettings.psd1')
```

Comment:
I'm not sure about this - in this case, the execution of Join-Path happens before the cmdlet is ever called, so this is just result of the `join-path` rather than an actual expression.

Reply:
I am not sure if it is really `Join-Path` because `(Join-path a b).GetType()` returns a string. As far as I understood when stepping through it in the debugger is that because we are in `BeginProcessing()`, the passed in object is not resolved to a string yet but the code was assuming that in the case of files that the input can be casted to a string in order to determine that it is of type 'file'.

Reply:
the place to look is in the setter of the parameter value. You could set a breakpoint in the setter (around line 215 of InvokeScriptAnalyzerCommand.cs) and inspect the value (or `Settings` after it's been set). The way powershell _should_ be doing this is because of the parens, that statement is evaluated _and then_ the parameter is bound to the string output of `join-path`. I haven't walked this through the debugger, but the way to check is that if Settings is the string result of the `join-path` (rather than the whole statement with parens) the check above should be definitive.

Reply:
You are right that PowerShell evaluates it in the setter but the object is of type `System.Management.Automation.PSObject`, whereas our property is just `System.Object` and therefore .Net truthfully says that it is not of type string. What I could do is try to cast it to `PSObject` then check whether its `BaseObject` property is of type string,

Reply:
if it winds up being not a string, that's good enough.